File: code_samples/section7/lesson/section7.cs
// ===========================================================
// Functions Under Test
// ===========================================================

/// <summary>
/// Counts how many times each word appears in the input sequence.
/// </summary>
/// <param name="words">
/// Sequence of words/tokens to count. (May be an array, list, LINQ query, etc.)
/// </param>
/// <returns>
/// A Dictionary mapping each distinct word to its frequency.
/// </returns>
/// <remarks>
/// Complexity:
///  - Time:  O(n) average, where n is the number of input words
///  - Space: O(k), where k is the number of distinct words
///
/// Notes:
///  - This implementation uses TryGetValue to avoid inserting missing keys until needed.
///  - Alternative simpler pattern in modern C#:
///      freq[w] = freq.GetValueOrDefault(w) + 1;   // (.NET 5+)
/// </remarks>
static Dictionary<string, int> CountWords(IEnumerable<string> words) {
    // Frequency table: word -> count
    var freq = new Dictionary<string, int>();

    // Iterate once through the sequence and update counts.
    foreach (var w in words) {
        // TryGetValue reads without adding a key.
        // If the key exists, we update it; otherwise we initialize it.
        if (freq.TryGetValue(w, out int value)) {
            // Pre-increment local copy (value) then store it back into the dictionary.
            // Example: value=2 -> ++value => 3, then freq[w]=3.
            freq[w] = ++value;
        } else {
            // First time we've seen this word.
            freq[w] = 1;
        }
    }

    return freq;
}

/// <summary>
/// Finds the first duplicate value encountered when scanning left-to-right.
/// </summary>
/// <param name="arr">Array of integers to scan.</param>
/// <returns>
/// The first duplicated integer, or null if all values are unique.
/// </returns>
/// <remarks>
/// Complexity:
///  - Time:  O(n) average
///  - Space: O(n) worst case (when all are unique)
///
/// Notes:
///  - HashSet.Add(x) returns false if x is already present.
///  - Returning int? avoids ambiguous sentinels (like 0 or -1).
/// </remarks>
static int? FirstDuplicate(int[] arr) {
    // Tracks values we've already seen as we scan the array.
    var seen = new HashSet<int>();

    foreach (var x in arr) {
        // Add returns false if x is already present => we found a duplicate.
        if (!seen.Add(x)) return x;
    }

    // No duplicates found.
    return null;
}

/// <summary>
/// Solves the "Two Sum" problem: find indices (i, j) such that nums[i] + nums[j] == target.
/// </summary>
/// <param name="nums">Input array of integers.</param>
/// <param name="target">Desired sum.</param>
/// <returns>
/// An int[] of length 2 containing the indices if found; otherwise [-1, -1].
/// </returns>
/// <remarks>
/// Approach:
///  - Maintain a dictionary value -> index for values seen so far.
///  - For each nums[i], compute need = target - nums[i].
///    If need was previously seen, we have a solution.
///
/// Complexity:
///  - Time:  O(n) average
///  - Space: O(n)
///
/// Notes / Edge cases:
///  - If the same value appears multiple times, indexOf[value] is overwritten with
///    the most recent index. This is fine for typical TwoSum requirements (any valid pair).
///  - Returns a C# collection expression [idx, i] (C# 12+) for the found case,
///    and new[] { -1, -1 } for the missing case.
/// </remarks>
static int[] TwoSum(int[] nums, int target) {
    // Map: number value -> index where that value was last seen.
    var indexOf = new Dictionary<int, int>(); // value -> index

    // Scan the array once.
    for (int i = 0; i < nums.Length; i++) {
        // Value needed to reach target with nums[i]
        int need = target - nums[i];

        // TryGetValue does lookup without inserting a new entry.
        if (indexOf.TryGetValue(need, out int idx)) {
            // Found a previous index 'idx' such that nums[idx] == need.
            // Return the two indices that form the target sum.
            return [idx, i];
        }

        // Record the current value's index for future matches.
        // If nums[i] repeats later, this overwrites with the newer index.
        indexOf[nums[i]] = i;
    }

    // No pair found that sums to target.
    return new[] { -1, -1 };
}

// ===========================================================
// Helpers
// ===========================================================

/// <summary>
/// Prints selected keys from a word-frequency dictionary in a readable format.
/// </summary>
/// <param name="label">Heading printed before the key/value lines.</param>
/// <param name="map">Dictionary to query.</param>
/// <param name="keys">Keys to display (some may not exist in the map).</param>
/// <remarks>
/// Why TryGetValue:
///  - Avoids inserting missing keys (unlike map[k] which would throw if missing).
/// </remarks>
static void PrintMap(string label, Dictionary<string, int> map, params string[] keys) {
    Console.WriteLine(label);

    foreach (var k in keys) {
        // Lookup without modifying the dictionary.
        if (map.TryGetValue(k, out int v))
            Console.WriteLine($"  '{k}' -> {v}");
        else
            Console.WriteLine($"  '{k}' -> (not found)");
    }
}

/// <summary>
/// Prints an integer array as: "Label[1, 2, 3]".
/// </summary>
/// <param name="label">Prefix before the formatted array.</param>
/// <param name="arr">Array to print.</param>
static void PrintArray(string label, int[] arr) {
    // string.Join formats the array elements separated by ", "
    Console.WriteLine(label + "[" + string.Join(", ", arr) + "]");
}

// ===========================================================
// TOP-LEVEL TESTS (no Main())
// ===========================================================
//
// This file uses top-level statements (C# 9+), so code here executes immediately
// when the program runs—no explicit Main() method is needed.
//

Console.WriteLine("=== Test: demoHash ===");

// Example dictionary initialization using collection initializer syntax.
// Here we pre-populate counts as if they were produced by counting logic.
var freq0 = new Dictionary<string, int>
{
    ["apple"] = 1,
    ["banana"] = 2
};

// Demonstrates safe read: TryGetValue returns false if key is missing.
// If missing, we use 0 as the fallback.
int n0 = freq0.TryGetValue("apple", out int value) ? value : 0;

Console.WriteLine($"apple count = {n0} (expected 1)\n");

// -----------------------------------------------------------

Console.WriteLine("=== Test: CountWords (word frequency) ===");

// Input words: apple x3, banana x2, orange x1
var words = new[] { "apple", "banana", "apple", "orange", "banana", "apple" };

// Run the function under test
var freq = CountWords(words);

// Print counts for a few keys, including one not present ("pear")
PrintMap("Frequencies:", freq, "apple", "banana", "orange", "pear");

Console.WriteLine("Expected: apple=3, banana=2, orange=1, pear not found\n");

// -----------------------------------------------------------

Console.WriteLine("=== Test: FirstDuplicate ===");

// Case 1: duplicates exist; the first repeated value encountered is 2
int[] arr1 = [2, 5, 1, 2, 3, 5, 1];
PrintArray("Array 1: ", arr1);

var dup1 = FirstDuplicate(arr1);
// dup1 is nullable; interpolating prints the value or blank if null.
Console.WriteLine($"First duplicate: {dup1} (expected 2)\n");

// Case 2: no duplicates; expect null
int[] arr2 = [1, 2, 3, 4, 5];
PrintArray("Array 2: ", arr2);

var dup2 = FirstDuplicate(arr2);
Console.WriteLine($"First duplicate: {dup2} (expected null)\n");

// -----------------------------------------------------------

Console.WriteLine("=== Test: TwoSum ===");

// Positive case: 2 + 7 = 9, expected indices (0,1) (order depends on implementation)
int[] nums = [2, 7, 11, 15];
int target = 9;

PrintArray("Array: ", nums);
Console.WriteLine("Target: " + target);

var pair1 = TwoSum(nums, target);
Console.WriteLine("twoSum result indices: [" + string.Join(", ", pair1) + "]");
Console.WriteLine("Expected: [0, 1] or [1, 0]\n");

// Negative test: no pair sums to 100 in {1,2,3}
Console.WriteLine("--- Negative case (no pair) ---");
int[] nums2 = [1, 2, 3];
int target2 = 100;

PrintArray("Array: ", nums2);
Console.WriteLine("Target: " + target2);

var pair2 = TwoSum(nums2, target2);
Console.WriteLine("twoSum result indices: [" + string.Join(", ", pair2) + "]");
Console.WriteLine("Expected: [-1, -1]\n");

File: code_samples/section7/lesson/section7.java
package code_samples.section7.lesson;

import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.Arrays;

/**
 * HashDemo
 *
 * Small standalone demo showing a basic HashMap lookup with a default value.
 *
 * Notes:
 * - HashMap is an unordered key/value store with average O(1) inserts/reads.
 * - getOrDefault(key, defaultValue) is a safe way to read a key that may not exist.
 */
class HashDemo {
    public static void main(String[] args) {
        // Map from word -> frequency (count)
        Map<String, Integer> freq = new HashMap<>();

        // Insert a couple of entries manually
        freq.put("apple", 1);
        freq.put("banana", 2);

        // Read with a fallback default (0) if the key isn't present
        int n = freq.getOrDefault("apple", 0);

        System.out.println("[HashDemo] apple count = " + n);
    }
}

/**
 * section7
 *
 * This class groups three classic hashing-based problems:
 *  1) countWords   - frequency counting using HashMap
 *  2) firstDuplicate - first repeated element detection using HashSet
 *  3) twoSum       - index-pair lookup using HashMap
 *
 * Educational intent:
 * - Emphasizes the "expected O(1)" lookup behavior of hash-based collections.
 * - Includes a simple test runner in main() to demonstrate behavior and outputs.
 *
 * Naming note:
 * - Java style conventions typically use PascalCase for class names (e.g., Section7).
 *   This file keeps your original name `section7` unchanged.
 */
public class section7 {

    /**
     * Counts the frequency of each word in the provided list.
     *
     * @param words List of strings (words/tokens) to count.
     * @return Map from word -> frequency.
     *
     * Complexity:
     * - Time:  O(n) average, where n is the number of words
     * - Space: O(k), where k is the number of distinct words
     *
     * Implementation details:
     * - getOrDefault(w, 0) returns 0 if the word has not been seen yet.
     * - We then store (existingCount + 1) back in the map.
     */
    Map<String, Integer> countWords(List<String> words) {
        // Frequency table: word -> count
        Map<String, Integer> freq = new HashMap<>();

        // Iterate over every word and increment its counter
        for (String w : words) {
            // If absent, treat as 0; otherwise, use current value and add 1
            freq.put(w, freq.getOrDefault(w, 0) + 1);
        }

        return freq;
    }

    /**
     * Finds the first duplicate value encountered when scanning left-to-right.
     *
     * Example:
     * - [2, 5, 1, 2, 3, 5, 1] -> returns 2
     *
     * @param arr Array of ints to scan.
     * @return The first duplicate, or null if none exists.
     *
     * Complexity:
     * - Time:  O(n) average
     * - Space: O(n) worst-case (when all values are unique)
     *
     * Notes:
     * - Returning Integer (nullable) avoids ambiguous sentinel values like 0 or -1.
     * - A more concise variant could be:
     *     if (!seen.add(x)) return x;
     *   because HashSet.add returns false if x was already present.
     */
    Integer firstDuplicate(int[] arr) {
        // Set of values we've already seen
        Set<Integer> seen = new HashSet<>();

        for (int x : arr) {
            // If x has been seen before, it's the first duplicate encountered so far
            if (seen.contains(x)) {
                return x;
            }
            // Otherwise record that we have now seen x
            seen.add(x);
        }

        // No duplicates found
        return null; // no duplicate
    }

    /**
     * Solves the classic Two Sum problem:
     * Find indices (i, j) such that nums[i] + nums[j] == target.
     *
     * @param nums Input array of integers.
     * @param target Desired sum.
     * @return int[2] containing the indices if found; otherwise [-1, -1].
     *
     * Complexity:
     * - Time:  O(n) average
     * - Space: O(n)
     *
     * Approach:
     * - Maintain a HashMap from number value -> index where it was last seen.
     * - For each nums[i], compute need = target - nums[i].
     *   If need exists in the map, we found a prior index j with nums[j] == need.
     *
     * Note about duplicates:
     * - indexOf.put(nums[i], i) overwrites previous indices for the same value.
     *   This is fine for typical Two Sum requirements (return any valid pair).
     */
    int[] twoSum(int[] nums, int target) {
        // Map: value -> index (where we last saw this value)
        Map<Integer, Integer> indexOf = new HashMap<>(); // value -> index

        for (int i = 0; i < nums.length; i++) {
            int need = target - nums[i];

            // Check if we've previously seen the needed value
            if (indexOf.containsKey(need)) {
                // Return the indices: previous index for need, and current index i
                return new int[] { indexOf.get(need), i };
            }

            // Record the current value with its index for future lookups
            indexOf.put(nums[i], i);
        }

        // No pair found
        return new int[] { -1, -1 };
    }

    // ===== Helpers for testing =====

    /**
     * Prints a subset of keys from a map, showing "(not found)" when missing.
     *
     * @param label Heading printed before the key/value lines.
     * @param map The map to inspect.
     * @param keys Varargs list of keys to attempt to print.
     *
     * Note:
     * - map.get(k) returns null if the key is missing.
     * - This works here because our values are Integer; however, note that a map could
     *   *also* store a key with a null value. If you need to distinguish "missing key"
     *   vs "present but null", use containsKey(k).
     */
    private static void printMap(String label, Map<String, Integer> map, String... keys) {
        System.out.println(label);

        for (String k : keys) {
            // get() returns the value, or null if the key doesn't exist
            Integer v = map.get(k);

            if (v != null) {
                System.out.println("  '" + k + "' -> " + v);
            } else {
                System.out.println("  '" + k + "' -> (not found)");
            }
        }
    }

    /**
     * Prints an integer array with a label using Arrays.toString().
     *
     * @param label Prefix to display before the array.
     * @param arr The array to print.
     */
    private static void printArray(String label, int[] arr) {
        System.out.println(label + Arrays.toString(arr));
    }

    // ===== Test runner =====

    /**
     * Main test harness.
     *
     * Runs three "tests" by printing:
     *  1) Frequency counts from countWords
     *  2) First duplicate from firstDuplicate (including a no-duplicate case)
     *  3) Two-sum indices from twoSum (including a no-solution case)
     *
     * This is not a formal unit test framework (like JUnit), but a simple
     * console-driven sanity check.
     */
    public static void main(String[] args) {
        // Create an instance to call the instance methods (countWords/firstDuplicate/twoSum)
        section7 s = new section7();

        // --- Test 1: countWords ---
        System.out.println("=== Test: countWords (word frequency) ===");

        // Build a sample list of words; expected counts:
        // apple=3, banana=2, orange=1
        List<String> words = Arrays.asList(
            "apple", "banana", "apple", "orange", "banana", "apple"
        );

        // Call the function under test
        Map<String, Integer> freq = s.countWords(words);

        // Print selected keys, including a missing one ("pear")
        printMap("Frequencies:", freq, "apple", "banana", "orange", "pear");
        System.out.println("Expected: apple=3, banana=2, orange=1, pear not found\n");

        // --- Test 2: firstDuplicate ---
        System.out.println("=== Test: firstDuplicate ===");

        // Case 1: duplicates exist; first duplicate encountered is 2
        int[] arr1 = {2, 5, 1, 2, 3, 5, 1};
        printArray("Array 1: ", arr1);

        Integer dup1 = s.firstDuplicate(arr1);
        System.out.println("First duplicate: " + dup1 + " (expected 2)\n");

        // Case 2: no duplicates; should return null
        int[] arr2 = {1, 2, 3, 4, 5};
        printArray("Array 2: ", arr2);

        Integer dup2 = s.firstDuplicate(arr2);
        System.out.println("First duplicate: " + dup2 + " (expected null = no duplicate)\n");

        // --- Test 3: twoSum ---
        System.out.println("=== Test: twoSum (hash map) ===");

        // Positive case: 2 + 7 = 9
        int[] nums = {2, 7, 11, 15};
        int target = 9;

        printArray("Array: ", nums);
        System.out.println("Target: " + target);

        int[] pair1 = s.twoSum(nums, target);
        System.out.println("twoSum result indices: " + Arrays.toString(pair1));

        if (pair1[0] != -1) {
            // Use returned indices to show the actual values used
            System.out.println("Values: " + nums[pair1[0]] + ", " + nums[pair1[1]]);
            System.out.println("Expected: indices [0,1] or [1,0] with values 2 and 7\n");
        } else {
            System.out.println("No pair found (unexpected for this test)\n");
        }

        // Negative test: no pair
        System.out.println("--- Negative case (no pair) ---");

        int[] nums2 = {1, 2, 3};
        int target2 = 100;

        printArray("Array: ", nums2);
        System.out.println("Target: " + target2);

        int[] pair2 = s.twoSum(nums2, target2);
        System.out.println("twoSum result indices: " + Arrays.toString(pair2));
        System.out.println("Expected: [-1, -1]\n");
    }
}
